### PR TITLE
Tigers - Neida Moreno

### PR DESCRIPTION
**Pull request by neidacreates**
9/27/22 - Turning this in after my partner and I finished it and passed all tests. I'm turning it in early in case I lose power because of hurricane. But my partner and I discussed refactoring some of the code to make it faster/nicer/shorter so if that happens it'll automatically update the pull request content?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -192,6 +192,10 @@ In addition to submitting your Pull Request, submit a reflection the the pair pr
 
 First, come up with a "plan of action" for how you want to work as a pair. Discuss your learning style, how you prefer to receive feedback, and one team communication skill you want to improve with this experience.
 
+Neida: Hurricane coming up in Florida - may lose access to internet. Learning style - reading and trying it on her own, drawing things out. Can accept feedback in anyway - perfers spoken feedback over written feedback and examples if its written. Asking for help earlier on. 
+
+Rahel: hands on learning learning, examples, visuals. Very open to feeback.
+
```
Comment (chimerror) [inline README.md:198]: Nice discussion!

File: adagrams/game.py [hunk 2]
```diff
@@ -1,11 +1,143 @@
+import random
 def draw_letters():
```
Comment (chimerror) [inline adagrams/game.py:2]: While this implementation functions, there are a few issues, including a logic error that makes it not match the requirements. I'll explain in other comments.

File: adagrams/game.py [hunk 3]
```diff
@@ -1,11 +1,143 @@
...
-    pass
+    '''
+    input: a list of letters 
+    output: a list of 10 letters of strings
+    '''
```
Comment (chimerror) [inline adagrams/game.py:6]: Good use of a docstring!

File: adagrams/game.py [hunk 4]
```diff
@@ -1,11 +1,143 @@
...
+    ''' 
+    # make letter pool which is a list of dictionaries
+    # make an editable copy of the letter pool
+    # set up an empty list that will hold the hand for the player
+    # loop through the range 1-10
+        # make a varaible that holds a random integer (1-26)
+        # go into the index of editable copy of list
+        # make a random integer between (1-the value) held within previous dictionary index
+        # append the letter however many times that random integer dictates
+    # return list
```
Comment (chimerror) [inline adagrams/game.py:15]: Pseudocode written comments like this can be incredibly helpful to organize your code, and I do like this style of pseudocode. However, even though this can be useful while scaffolding, I recommend removing it from the code once you have the actual code down.

This type of comment, as a comment is what I call a "what" comment. It says _what_ the code does. However, while it may _say_ one thing, the actual thing that the code does is determined by the _code_. So for example, these comments:

```python
        # make a random integer between (1-the value) held within previous dictionary index
        # append the letter however many times that random integer dictates
```

do not match what the code actually _does_ in that section:

```python
        key = list(letter_dict.keys())[0]
        val = list(letter_dict.values())[0]

        if val != 0: 
            hand_list.append(key)
            letter_dict[key] -= 1 
```

My guess is that y'all wrote the pseudocode when y'all had a different idea for the implementation, and then changed the implementation to something else when it didn't work as expected. Which is fine! But that's also the problem with "what" comments, they tend to fall out of sync with the code they are describing.

Even if they are not out of sync, they are still redundant, and thus repeating yourself, and thus the type of thing you want to avoid when writing code. Instead I recommend trying to keep your comments to only be "why" comments.

Rather than being about _what_ the code does, they are about _why_ the code does it the way it does. A lot of time this is something like `# Using slice to make a copy of the list so it doesn't get modified". It explains why code may be performing a step that seems counter to what a reader would expect. _Why_ a code is a certain way is often much more useful than a repetition of _what_ the code does.

If it is hard to tell what the code does from the code, take that as a sign that there is likely a better way to write the code so that it _is_ easier to understand.

File: adagrams/game.py [hunk 5]
```diff
@@ -1,11 +1,143 @@
...
+        {"V" : 2}, 
+        {"J" : 1}, 
+        {"W" : 2}, 
+        {"K" : 1}, 
+        {"X" : 1}, 
+        {"L" : 4}, 
+        {"Y" : 2}, 
+        {"M" : 2}, 
+        {"Z" : 1}
+    ]
```
Comment (chimerror) [inline adagrams/game.py:44]: This is a very odd choice for a data structure here. Usually this type of information would be represented as a single dictionary with a key-value pair for each letter:

```python
letter_pool = {
    "A":  9,
    "B": 2,
    "C": 2,
     # and so on...
}
```

That is, there's not much benefit to a bunch of little dictionaries compared to just one dictionary. My guess is that y'all wanted a way to pick one of the keys at random, and thus, chose instead to wrap them in the list so y'all could use `random.randint` to generate a random index. In that case, it may have been easier to pull out only the keys into a separate variable:

```python
LETTERS = list(letter_pool.keys())
```

And then we can get a random letter later on with:

```python
rand_num = random.randint(0, 25)
rand_letter = LETTERS[rand_num]
```

But we can get even better with a different function in the `random` module, `choice`:

```python
rand_letter = random.choice(LETTERS)
```

Of course, this is going to change how the rest of the code needs to be written, and I'll talk more about that in the next comment.

File: adagrams/game.py [hunk 6]
```diff
@@ -1,11 +1,143 @@
...
+    letter_pool_copy = letter_pool[:]
+    hand_list = []
+
+    while len(hand_list) < 10:
+        rand_num = random.randint(0, 25)
+        letter_dict = letter_pool_copy[rand_num] 
+        key = list(letter_dict.keys())[0]
+        val = list(letter_dict.values())[0]
```
Comment (chimerror) [inline adagrams/game.py:52]: The real benefit of my suggestion for `letter_pool` being a single dictionary shows up here, as y'all's rather awkward code to access the tiny dictionary can instead become:

```python
rand_letter = random.choice(LETTERS)
letters_left = letter_pool[rand_letter]

if letters_left > 0:
    hand_list.append(rand_letter)
    letter_pool[rand_letter] -= 1
```

Which is much more readable. Y'all may notice that I'm working with `letter_pool` directly, which brings me to another comment.

File: adagrams/game.py [hunk 7]
```diff
@@ -1,11 +1,143 @@
...
+        {"J" : 1}, 
+        {"W" : 2}, 
+        {"K" : 1}, 
+        {"X" : 1}, 
+        {"L" : 4}, 
+        {"Y" : 2}, 
+        {"M" : 2}, 
+        {"Z" : 1}
+    ]
+    letter_pool_copy = letter_pool[:]
```
Comment (chimerror) [inline adagrams/game.py:45]: There's not really much reason to make a copy here from what I can see. `letter_pool` is not re-used outside of this function, and it gets reset to the starting values every time the function is recalled any way.

But let's round this out with the logic error...

File: adagrams/game.py [hunk 8]
```diff
@@ -1,11 +1,143 @@
...
+    hand_list = []
+
+    while len(hand_list) < 10:
+        rand_num = random.randint(0, 25)
```
Comment (chimerror) [inline adagrams/game.py:49]: By pulling y'all's random choice from the letters, there's a slight difference in the random distribution of letters from the requirements, that is the likelihood of each letter being drawn is off (which is a logic error). For example, since there are 12 "E" tiles and 6 "N" tiles, it should be twice as likely to draw an "E" over an "N". However, as each letter shows up only once, they would all have equal probability of being drawn here (as would every other letter).

Y'all's code that is below this will keep y'all from drawing too many of the letters, but it can't fix the distribution.

Y'all would have to figure out a solution to match the desired distribution on y'all's own, but let me summarize my other comments into a cleaner implementation of y'all's algorithm.

File: adagrams/game.py (hunk 5 above)
Comment (chimerror) [reply adagrams/game.py:44]: By the way, the reason I capitalized `LETTERS` is because the list of letters is not going to change, and thus it's a constant and writing it in all-caps is the Python convention for representing that.

File: adagrams/game.py [hunk 9]
```diff
@@ -1,11 +1,143 @@
...
+        letter_dict = letter_pool_copy[rand_num] 
+        key = list(letter_dict.keys())[0]
+        val = list(letter_dict.values())[0]
+
+        if val != 0: 
+            hand_list.append(key)
+            letter_dict[key] -= 1 
+    return hand_list
```
Comment (chimerror) [inline adagrams/game.py:57]: Putting all my suggestions together would give us this:

```python
def draw_letters():
    letter_pool = {
        "A":  9,
        "B": 2,
        "C": 2,
        # and so on...
    }
    LETTERS = list(letter_pool.keys())
    hand_list = []

    while len(hand_list) < 10:
        rand_letter = random.choice(LETTERS)
        letters_left = letter_pool[rand_letter]

        if letters_left > 0:
            hand_list.append(rand_letter)
            letter_pool[rand_letter] -= 1

    return hand_list
```

Which still has the same logic error and incorrect random distribution of y'all's code, but feels less awkward and easier to read.

File: adagrams/game.py [hunk 10]
```diff
@@ -1,11 +1,143 @@
...
+
 
 def uses_available_letters(word, letter_bank):
```
Comment (chimerror) [inline adagrams/game.py:60]: This looks fine, just a couple of minor comments.

File: adagrams/game.py [hunk 11]
```diff
@@ -1,11 +1,143 @@
...
+    available_letters = []
+    letter_bank_copy = letter_bank[:]
+    for char in word.upper():
+        if char not in letter_bank_copy:
+            return False
+        else:
+            available_letters.append(char)
+            letter_bank_copy.remove(char)
+            if len(available_letters) == len(word):
+                return True
```
Comment (chimerror) [inline adagrams/game.py:76]: By having this `if len(available_letters) == len(word):` block at this indentation level, y'all're performing this check on every iteration of the outer `for` loop that hits the `else` block, which doesn't seem to have much benefit compared to checking it outside of the loop:

```python
    for char in word.upper():
        if char not in letter_bank_copy:
            return False
        else:
            available_letters.append(char)
            letter_bank_copy.remove(char)
    if len(available_letters) == len(word):
        return True
```

That is, there is no possible way that `len(available_letters) == len(word)` could evaluate to `True` before we have gone over every letter in `word.upper()`. Let's say there are N letters in `word`. Each iteration we add possibly 1 letter at most to `available_letters`. So since `len(available_letters)` starts at 0 it can't be N until we've looked at the N letters of `word`.

It's a very minor performance hit to do this needless checking, and usually something like this wouldn't kill your performance, but as your code begins working with things that may take some time to run (like, say, API calls) that can begin to add up.

File: adagrams/game.py [hunk 12]
```diff
@@ -1,11 +1,143 @@
...
 
 def uses_available_letters(word, letter_bank):
-    pass
+    #loop through the letters in the word and check if its in the letter bank
+    #if its not in the letter bank, return false
+    #potentially write a helper function
+    #initialize a list to hold all present letters from the letter_bank 
+    #if the letter is in the bank, add it to the new list and remove it from the letter bank
+    #if the len of new list == len word return true else false 
+    available_letters = []
```
Comment (chimerror) [inline adagrams/game.py:67]: This name feels a bit backwards, because the "available" letters are actually in `letter_bank`, so this variable would make more sense to be named something like `used_letters`.

File: adagrams/game.py [hunk 13]
```diff
@@ -1,11 +1,143 @@
...
+        if char not in letter_bank_copy:
+            return False
+        else:
+            available_letters.append(char)
+            letter_bank_copy.remove(char)
+            if len(available_letters) == len(word):
+                return True
+
 
 def score_word(word):
```
Comment (chimerror) [inline adagrams/game.py:79]: This works, but much like `draw_letters`, the choice of data structure here makes the code a bit more awkward and causes a performance hit.

File: adagrams/game.py [hunk 14]
```diff
@@ -1,11 +1,143 @@
...
+    # if len of word >= 7 add 8 more points
+    # return score 
+
+    list_1 = ["A", "E", "I", "O", "U", "L", "N", "R", "S", "T"]
+    list_2 = ["D", "G"]
+    list_3 = ["B", "C", "M", "P"]
+    list_4 = ["F", "H", "V", "W", "Y"]
+    list_5 = ["K"]
+    list_8 = ["J", "X"]
+    list_10 = ["Q", "Z"]
```
Comment (chimerror) [inline adagrams/game.py:92]: There is an aspect of this list of variables that I like, where it clearly avoids repeating the number of points each letter is worth. However despite that, I think a dictionary from the letter to the points would work better:

```python
POINTS = {
    "A" : 1,
    "B": 3,
    "C": 3,
    "D": 2,
    # and so on...
}
```

Yes, this is a lot of repetition, but there will be a performance benefit that I'll talk about in a comment below.

Comment (chimerror) [reply adagrams/game.py:92]: There is actually a pretty cool way that I noticed while grading that can reduce the repetition, but it's kind of ugly and using `dict.fromkeys` and the `**` dictionary unpacking operator:

```python
POINTS = {
      **dict.fromkeys(["A", "E", "I", "O", "U", "L", "N", "R", "S", "T"], 1),
      **dict.fromkeys(["D", "G"], 2),
      **dict.fromkeys(["B", "C", "M", "P"], 3),
      **dict.fromkeys(["F", "H", "V", "W", "Y"], 4),
    "K" : 5,
      **dict.fromkeys(["J", "X"], 8),
      **dict.fromkeys(["Q", "Z"], 10)
    }
```

But this is pretty arcane, so I would not be surprised to see just the flat dictionary like I suggested. Especially in student code, but I wouldn't give a flat dictionary a second thought in industry code either.

(Y'all can read more about `dict.fromkeys` [here in the Python documentation](https://docs.python.org/3/library/stdtypes.html#dict.fromkeys) and the `**` as a method for merging dictionaries [here on GeeksForGeeks](https://www.geeksforgeeks.org/python-merging-two-dictionaries/).)

Comment (chimerror) [reply adagrams/game.py:92]: Note once again, I capitalized `POINTS` because it will not change and thus is a constant.

File: adagrams/game.py [hunk 15]
```diff
@@ -1,11 +1,143 @@
...
+        elif char in list_3:
+            score += 3
+        elif char in list_4:
+            score += 4
+        elif char in list_5:
+            score += 5
+        elif char in list_8:
+            score += 8
+        elif char in list_10:
+            score += 10
```
Comment (chimerror) [inline adagrams/game.py:110]: The main benefit to a flat dictionary is that because the keys of `POINTS` are the letters themselves so this entire `if`-`elif` block can be replaced with direct access of the `POINTS` dictionary:

```python
for char in word.upper():
    score += POINTS[char]
```

There is another benefit with my implementation besides less code, namely that it avoids a performance issue. The `in` operator on lists works by comparing each element in the list in order, which is a linear O(n) time complexity operation where n is the number of elements in the list. So for example, if `char` was `"Z"`, then we will check all the letters in `list_1`, and then in `list_2`, and then in `list_3` and so on until we get to the very end of `list_10` where we find `"Z"`.

However, to access a value from a dictionary using a key uses a hash function to look up the corresponding value from the key, which takes constant O(1) time complexity. (There's a caveat to this I'll explain in another comment.)

Basically think about it like this: `in` on a list is like starting at the front of a paper dictionary and going through each word until you find the one you're looking for. On the other hand, accessing a dictionary by its keys (`my_dict[key]`) is like going to that word directly by using the little markers for each letter on the side as well as the little words at the top that tell you which words are on that page. (By the way, this rather outdated reference to paper dictionaries is exactly why this data structure is called a "dictionary".)

Comment (chimerror) [reply adagrams/game.py:110]: My caveat about the time complexity of accessing a dictionary follows; feel free to skip reading this. 

There will be more discussion about hashing in a CS fundamentals lesson later I believe, but the way it basically works is by taking the desired key and using a function to generate a number called a hash, and using that to look up that key. However, it then becomes possible that the hash value for different keys are the same, what's called a hash collision. In this case, the values of those colliding keys will basically be added under that same hash in a list.

If, in the worst case, every key collides, all the values will be in the same list under the hash and thus we lose all the magic of the hash, and it becomes a list again. Thus, the worst-case time complexity of accessing a dictionary by key is O(n).

_However_, with a good hash function (which I feel pretty certain the one Python uses for its dictionaries is) collisions should be so exceedingly rare that we can just use the most likely time complexity of O(1).

File: adagrams/game.py [hunk 16]
```diff
@@ -1,11 +1,143 @@
...
+    if len(word) >= 7:
+        score += 8
+    return score
 
 def get_highest_word_score(word_list):
```
Comment (chimerror) [inline adagrams/game.py:115]: This looks fine, but there's a slight improvement that can be made, as well as a small naming suggestion.

File: adagrams/game.py [hunk 17]
```diff
@@ -1,11 +1,143 @@
...
+    #loop through the word_list and for each word we call the function above to get score
+        # compare it to the higheset_score_var and it's replaced if it is higher
+        # add the highest score word to the string var
+        # as we loop through replace them using conditionals
+            # if tie compare len of words with len of highest word place holder (if its less than highest word it replaces current one)
+                # if len(word) > len(string_var) and len(strint_var) is not 10 then replace with new word
+                #elif len(word) == len(string_var) then string_var will remain
+        # return tuple 
+    highest_score = 0
+    highest_str = ""
```
Comment (chimerror) [inline adagrams/game.py:127]: In general, I don't recommend naming variables after their types like this. It's usually more informative to describe what the variable represents rather than what type it is. As such I would prefer a name like `highest_word` rather than `highest_str`.

File: adagrams/game.py [hunk 18]
```diff
@@ -1,11 +1,143 @@
...
+    for word in word_list:
+        word_score = score_word(word)
+        if word_score > highest_score:
+            highest_score = word_score 
+            highest_str = word
+        elif word_score == highest_score:
+            if len(word) < len(highest_str) and len(highest_str) != 10:
+                highest_str = word
+            elif len(word) == 10 and len(highest_str) != 10:
+                highest_str = word
```
Comment (chimerror) [inline adagrams/game.py:138]: This loop works fine, but I would like to show some improvements that can be had by trying to make the code more DRY by removing repetition.

Let's start with all these `len` calls. Even though they take constant O(1) time on lists and strings, I think it's helpful to avoid having to call them multiple times, so let's make a few helper variables (I'll keep using `highest_str` for now, just so it's clear the development from y'all's code):

```python
highest_length = len(highest_str)
word_length = len(word)
if word_length < highest_length and highest_length != 10:
    highest_str = word
elif word_length == 10 and highest_length != 10:
    highest_str = word
```

Now the next thing I notice is how `and highest_length != 10` is repeated twice, so by pulling it out into its own `if` block, I can not just remove repetition, but also move the `word_length = len(word)` assignment in there, so we don't have to call `len(word)` unless we need to:

```python
highest_length = len(highest_str)
if highest_length != 10:
    word_length = len(word)
    if word_length < highest_length:
        highest_str = word
    elif word_length == 10:
        highest_str = word
```

The next thing I notice is that the body of the inner `if` and `elif` blocks is exactly the same: `highest_str = word`. I can combine them with the `or` operator:

```python
highest_length = len(highest_str)
if highest_length != 10:
    word_length = len(word)
    if word_length < highest_length or word_length == 10:
        highest_str = word
```

And there, no repetition!

I wanted to show a progression from y'all's code because I definitely know that I often end up writing code like that as I struggle to figure things out, and wanted to show a way to then work from that code to clean up the repetition. You don't have to write it the cleanest way the first time! It can be easier to get it to work, and then clean up afterwards.

**Review by chimerror — COMMENTED** [review]
Almost there!

While this code does work fine and pass all of our tests except for a logic issue with the probability of `draw_letters`, I do think there's some awkward decisions on choices of data structure that has me a bit worried about how y'all's knowledge of the material, so I am marking this Yellow.

That being said, Yellow is an invitation to take another pass at this to really make sure y'all understand the material. However, note that I say "invitation"; y'all are _not_ required to resubmit this, I just feel it may be good practice to gain familiarity with data structures. But y'all do not need to do so, especially with how late I got this grade in, since it's been a few weeks and y'all are likely much more practiced than y'all were when y'all wrote this.

If y'all do decide to resubmit, I've given a lot of suggestions for more appropriate data structures, so it shouldn't be too much work to update it.
